### PR TITLE
Enable health checks in kubernetes

### LINKED ISSUE #953: Support k8s custom pod health checks
Change #934 disabled passthrough ports for health checks. We need to re-introduce the support back.

### PR DESCRIPTION
**Pull request by rshriram**
Fixes #953

### CODE REVIEW

File: model/service.go [hunk 1]
```diff
@@ -201,6 +201,14 @@ type ServiceDiscovery interface {
 
 	// HostInstances lists service instances for a given set of IPv4 addresses.
 	HostInstances(addrs map[string]bool) []*ServiceInstance
+
+	// ManagementPorts lists set of management ports associated with an IPv4 address.
```
Comment (kyessenov) [inline model/service.go]: Might be worth explaining what purpose these ports serve in istio context: the proxy should allow connections through this port for transparent proxy injection.

File: platform/kube/conversion.go [hunk 2]
```diff
@@ -18,10 +18,14 @@ import (
 	"fmt"
 	"strings"
```
Comment (kyessenov) [inline platform/kube/conversion.go]: remove whitespace

File: platform/kube/conversion.go [hunk 3]
```diff
@@ -124,3 +128,104 @@ func convertProtocol(name string, proto v1.Protocol) model.Protocol {
...
+					}
+					// Deduplicate along the way. We don't differentiate between HTTP vs TCP mgmt ports
+					if set[p.Name] == nil {
+						set[p.Name] = p
+					}
+				}
+			}
+		}
+
+		if container.ReadinessProbe != nil {
```
Comment (kyessenov) [inline platform/kube/conversion.go]: please de-duplicate the code (HTTPGetAction in Probe is same for liveness and readiness)

File: model/service.go (hunk 1 above)
Comment (rshriram) [inline model/service.go]: Its already there in the explanation.

File: platform/kube/conversion_test.go [hunk 4]
```diff
@@ -201,3 +202,97 @@ func TestInvalidExternalServiceConversion(t *testing.T) {
...
+
+			podSpec.Containers[0].LivenessProbe.Handler = handler1
+			podSpec.Containers[0].ReadinessProbe.Handler = handler2
+
+			mgmtPorts, err := convertProbesToPorts(podSpec)
+			if err != nil {
+				t.Errorf("Failed to convert Probes to Ports: %v", err)
+			}
+
+			if len(mgmtPorts) != len(expected) {
```
Comment (kyessenov) [inline platform/kube/conversion_test.go]: nit: could use reflect.DeepEqual here

File: proxy/envoy/config.go [hunk 5]
```diff
@@ -162,6 +164,12 @@ func buildListeners(env proxy.Environment, sidecar proxy.Sidecar) (Listeners, Cl
 		insertMixerFilter(listeners, instances, sidecar)
 	}
 
+	// Append management listeners after injecting mixer filter to normal listeners
+	// because management port listeners cannot use mixer or auth (see comment
+	// under buildMgmtPortListeners)
+	listeners = append(listeners, mgmtListeners...)
```
Comment (kyessenov) [inline proxy/envoy/config.go]: The previous iteration would de-duplicate management and inbound listeners. Can we at least skip management listeners that are also inbound listeners? This used to cause segfaults in Envoy before LDS.

File: proxy/envoy/config.go [hunk 6]
```diff
@@ -432,46 +488,13 @@ func buildInboundListeners(instances []*model.ServiceInstance,
...
-
-		case model.ProtocolTCP, model.ProtocolHTTPS:
-			listeners = append(listeners, buildTCPListener(&TCPRouteConfig{
-				Routes: []*TCPRoute{buildTCPRoute(cluster, []string{endpoint.Address})},
-			}, endpoint.Address, endpoint.Port))
 
-		default:
-			glog.Warningf("Unsupported inbound protocol %v for port %#v", protocol, servicePort)
+		listener, cluster := buildInboundConfig(endpoint.Address, endpoint.Port, servicePort.Protocol, mesh)
+		if cluster != nil && listener != nil {
```
Comment (kyessenov) [inline proxy/envoy/config.go]: refactor to use errors instead of nils

File: proxy/envoy/config.go [hunk 7]
```diff
@@ -481,3 +504,38 @@ func buildInboundListeners(instances []*model.ServiceInstance,
...
+// that the health check ports are distinct from the service ports.
+func buildMgmtPortListeners(managementPorts model.PortList,
+	managementIP string, mesh *proxyconfig.ProxyMeshConfig) (Listeners, Clusters) {
+	listeners := make(Listeners, 0, len(managementPorts))
+	clusters := make(Clusters, 0, len(managementPorts))
+
+	// assumes that inbound connections/requests are sent to the endpoint address
+	for _, mPort := range managementPorts {
+		listener, cluster := buildInboundConfig(managementIP, mPort.Port, mPort.Protocol, mesh)
+		if cluster != nil && listener != nil {
```
Comment (kyessenov) [inline proxy/envoy/config.go]: same, use errors instead

File: proxy/envoy/config.go (hunk 5 above)
Comment (rshriram) [inline proxy/envoy/config.go]: Not sure I understand. the normalize() calls are still happening before returning the final set of listeners and clusters. I thought the normalize was responsible to deduplicate

Comment (kyessenov) [inline proxy/envoy/config.go]: normalize assumes that the duplicates are identical. otherwise, we get non-determinism which one gets chosen.

**Comment by rshriram** [comment]
I added the health check port to the integration test YAML. If auth was applied, then the e2e test wouldn't pass right?

**Comment by rshriram** [comment]
@kyessenov PTAL. Updated goldens

File: proxy/envoy/config.go [hunk 8]
```diff
@@ -183,13 +156,29 @@ func buildClusters(env proxy.Environment, role proxy.Node) (clusters Clusters) {
...
+	for _, m := range mgmtListeners {
+		l := listeners.GetByAddress(m.Address)
+		if l != nil {
+			glog.Warningf("Omitting listener for management address %s (%s) due to collision with service listener %s (%s)",
+				m.Name, m.Address, l.Name, l.Address)
+			continue
+		}
+		listeners = append(listeners, m)
+	}
+	clusters = append(clusters, mgmtClusters...)
```
Comment (kyessenov) [inline proxy/envoy/config.go]: we shouldn't be adding clusters for non-existent listeners, they might override clusters for non-management listeners

**Comment by kyessenov** [comment]
Are you sure goldens are right? Doesn't look like they match the code.

**Pull request merged**